lib/activedirectory/protocol/krb5.c

```diff
@@ -409,17 +430,67 @@ static PyMethodDef k5_methods[] =
 };
 
 
+#if PY_MAJOR_VERSION >= 3
+
+static int k5_traverse(PyObject *m, visitproc visit, void *arg) {
+    Py_VISIT(GETSTATE(m)->error);
+    return 0;
+}
+
+static int k5_clear(PyObject *m) {
+    Py_CLEAR(GETSTATE(m)->error);
+    return 0;
+}
+
+
+static struct PyModuleDef moduledef = {
+        PyModuleDef_HEAD_INIT,
+        "krb5",
+        NULL,
+        sizeof(struct module_state),
+        k5_methods,
+        NULL,
+        k5_traverse,
+        k5_clear,
+        NULL
+};
+
+#define INITERROR return NULL
+
+PyMODINIT_FUNC
+PyInit_krb5(void)
+
+#else
+#define INITERROR return
+
 void
 initkrb5(void)
+#endif
 {
     PyObject *module, *dict;
 
 #if !defined(__APPLE__) || !defined(__MACH__)
     initialize_krb5_error_table();
 #endif
 
+#if PY_MAJOR_VERSION >= 3
+    module = PyModule_Create(&moduledef);
+#else
     module = Py_InitModule("krb5", k5_methods);
+#endif
+
+    if (module == NULL)
+        INITERROR;
+
     dict = PyModule_GetDict(module);
-    k5_error = PyErr_NewException("freeadi.protocol.krb5.Error", NULL, NULL);
-    PyDict_SetItemString(dict, "Error", k5_error);
+
+    struct module_state *st = GETSTATE(module);
+    st->error = PyErr_NewException("freeadi.protocol.krb5.Error", NULL, NULL);
```

Comment:
You'll still need to set `k5_error` so that the other methods have a reference to the exception, e.g.:

```c
k5_error = st->error;
```

Otherwise if you perform an action that would trigger an exception from the C code, the `PyErr_Format` calls don't go through and you get `SystemError: <built-in method change_password ...> returned NULL without setting an error`. Here's a relatively simple test case

```python
import activedirectory

AD_SITE = 'Redacted'
BIND_USER = 'redacted'
BIND_PASSWORD = 'redacted'
DOMAIN = 'redacted.com'

creds = activedirectory.Creds(DOMAIN)
creds.acquire(BIND_USER, BIND_PASSWORD)
activedirectory.activate(creds)
client = activedirectory.Client(DOMAIN, creds)

# Should raise instance of activedirectory.core.exception.Error with message:
# 'krb5_get_init_creds_password(): Client not found in Kerberos database'
client.change_password('baduser@%s' % DOMAIN, 'a', 'a')
```